rest/oidc_api_test.go

```diff
@@ -863,3 +870,266 @@ func assertHttpResponse(t *testing.T, response *http.Response, statusCode int, b
+					CallbackStateCookieHTTPOnly: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+		}, {
+			name: "mitigation from csrf when callback state enabled",
+			inputProviders: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+			invalidateState:      true,
+		}, {
+			name: "vulnerable to csrf when callback state disabled",
+			inputProviders: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("quz"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+			invalidateState:      true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			opts := auth.OIDCOptions{Providers: tc.inputProviders, DefaultProvider: &tc.inputDefaultProvider}
+			restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &opts}}
+			restTester := NewRestTester(t, &restTesterConfig)
+			defer restTester.Close()
+			mockSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+			defer mockSyncGateway.Close()
+			mockSyncGatewayURL := mockSyncGateway.URL
+
+			if tc.invalidateState {
+				mockAuthServer.SetOption("invalidateState", true)
+				defer mockAuthServer.DeleteOption("invalidateState")
+			}
+
+			// Initiate OpenID Connect Authorization Code flow.
+			requestURL := mockSyncGatewayURL + tc.inputRequestURL
+			request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+			require.NoError(t, err, "Error creating new request")
+			jar, err := cookiejar.New(nil)
+			require.NoError(t, err, "Error creating new cookie jar")
+			client := &http.Client{Jar: jar}
+			response, err := client.Do(request)
+			require.NoError(t, err, "Error sending request")
+			defer func() { require.NoError(t, response.Body.Close(), "Error closing response body") }()
+
+			if tc.invalidateState {
+				bodyBytes, err := ioutil.ReadAll(response.Body)
+				require.NoError(t, err, "error reading response body")
+				assert.Contains(t, string(bodyBytes), "State mismatch")
+				assert.Equal(t, http.StatusBadRequest, response.StatusCode)
+				return
+			}
```

Comment:
If this is a special case, it might make sense to move this into a separate test to simplify the existing set of scenarios.

Reply:
This change has been revised in latest revision.